### DISCUSSION TITLE
Calculate network motif

### DISCUSSION THREAD

**Discussion by visunf** [networkx/networkx#7218] [Algorithms]
I hope networkx can add a function to calculate motifs. Even though igraph has a function to count the number of motifs, there is no function to return a list of motifs.
Specifically,

input: graph, n   # n: n-node motifs
ouput: [{subgraph_1, subgraph_2, subgraph_3....}] # return all n-nodes motifs

```python
def getMotifs(graph, n):
       .....
      return [subgraph_1, subgraph_2, subgraph_3, ....] # return all n-nodes motifs
```
Learn from 
what is network motif: 
`https://en.wikipedia.org/wiki/Network_motif`
other python package about motif:
`https://python.igraph.org/en/stable/api/igraph.GraphBase.html#motifs_randesu`
and
`Milo R, Shen-Orr S, Itzkovitz S, et al. Network motifs: simple building blocks of complex networks[J]. Science, 2002, 298(5594): 824-827.`
`Reference: S. Wernicke and F. Rasche: FANMOD: a tool for fast network motif detection, Bioinformatics 22(9), 1152--1153, 2006.`

**Comment by dschult** [comment]
NX currently has functions for making a triad census (n=3). 
We'd be quite interested in other methods. We do have the `subgraph_monomorphism` and isomorphism tools as well which is where I would expect users with an interest in this functionality to go.

The wikipedia article sited above has many nice algorithms for motif reporting. 

Let's use this space to discuss what would be of interest and how to get that implemented. :}

**Comment by szhorvat** [comment]
> Even though igraph has a function to count the number of motifs, there is no function to return a list of motifs.

This sounds like a complaint about igraph, so you might consider bringing it up there as well 😉 

But in fact python-igraph _does_ support enumerating individual induced subgraphs. See the `callback` parameter of `motifs_randesu()`. You can implement a callback that collects the subgraphs. 

Furthermore, the underlying C library's motif finder is flexible enough to easily implement additional efficient motif analyses as well, [as I have done in the Mathematica interface of igraph](http://szhorvat.net/mathematica/IGDocumentation/#igmotifsvertexparticipation). Thus if you present a good argument for a specific improvement, it might be fairly easy to add it to the Python library.
